File: AutoPecas.Service/MarcaService.cs
using Autopecas.Infra.Data;
using AutoPecas.Core.Model;
using AutoPecas.Core.Spec;
using LinqKit;
using Microsoft.EntityFrameworkCore;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;

namespace AutoPecas.Service
{
    public class MarcaService
    {
        private readonly AutoPecasDbContext _AutoPecasDbContext;

        private void AplicarFiltro(FiltroSpec filtro, out IQueryable<Marca> query)
        {
            query = _AutoPecasDbContext.Marcas
                        .AsExpandableEFCore();

            var predicate = PredicateBuilder.New<Marca>(true);

            if (filtro.Filtros.TryGetValue("nome", out var nome))
            {
                predicate.And(p => p.Nome.Contains((string)nome));
            }

            if (filtro.Filtros.TryGetValue("descricao", out var descricao))
            {
                predicate.Or(p => p.Descricao.Contains((string)descricao));
            }

            query = query.Where(predicate);
        }

        public MarcaService(AutoPecasDbContext autoPecasDbContext)
        {
            _AutoPecasDbContext = autoPecasDbContext;
        }

        public async Task<PaginacaoResultado<Marca>> Lista(FiltroSpec filtro)
        {
            AplicarFiltro(filtro, out var query);

            var resultado = new PaginacaoResultado<Marca>(query, filtro.Pagina, filtro.Tamanho, filtro.Total);

            resultado.Lista = await query
                .Skip(filtro.Tamanho * (filtro.Pagina - 1))
                .Take(filtro.Tamanho)
                .ToListAsync();

            return resultado;
        }

        public async Task<Marca> Obter(int idMarca)
        {
            return await _AutoPecasDbContext
                .Marcas
                .Where(p => p.Id == idMarca)
                .FirstOrDefaultAsync();
        }

        public async Task<int> Incluir(Marca marca)
        {
            _AutoPecasDbContext.Add(marca);

            return await _AutoPecasDbContext.SaveChangesAsync();
        }

        public async Task<IList<Marca>> Busca(string texto)
        {
            return await _AutoPecasDbContext
                .Marcas
                .Where(p => p.Nome.Contains(texto))
                .Take(5)
                .ToListAsync();
        }
        public async Task<int> Editar(Marca marca)
        {
            _AutoPecasDbContext.Update(marca);
            return await _AutoPecasDbContext.SaveChangesAsync();
        }
    }
}

File: AutoPecas.Web/ClientApp/src/app/service/marca.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { FiltroSpec } from '../model/geral/filtro-spec.model';
import { PaginacaoResultado } from '../model/geral/paginacao-resultado.model';
import { Marca } from '../model/produto/marca.model';
import { environment } from 'src/environments/environment';

@Injectable({
  providedIn: 'root',
})

export class MarcaService {
  private readonly apiEndpoint = `${environment.apiUrl}marca`;

  constructor(private httpClient: HttpClient) { }

  public listar(filtro: FiltroSpec) {
    return this.httpClient.post<PaginacaoResultado<Marca>>(`${this.apiEndpoint}/lista/`, filtro);
  }

  public busca(texto: string): Observable<Marca[]> {
    return this.httpClient.get<Marca[]>(`${this.apiEndpoint}/busca/${texto}`);
  }

  public obter(id: number): Observable<Marca> {
    return this.httpClient.get<Marca>(`${this.apiEndpoint}/${id}`);
  }

  public incluir(marca: Marca): Observable<number> {
    return this.httpClient.post<number>(`${this.apiEndpoint}`, marca);
  }

  public editar(marca: Marca): Observable<number> {
    return this.httpClient.put<number>(`${this.apiEndpoint}`, marca);
  }
}
